File: api/workflows.py
from __future__ import annotations

from fastapi import APIRouter, HTTPException
from pydantic import BaseModel, Field

from application.workflows import (
    cancel_workflow_batch,
    cancel_workflow_run,
    create_workflow_batch,
    create_or_update_workflow_definition,
    create_workflow_run,
    get_workflow_batch,
    get_workflow_batch_summary,
    get_workflow_run,
    get_workflow_run_summary,
    list_workflow_adapters,
    list_workflow_batches,
    list_workflow_definitions,
    list_workflow_events,
    list_workflow_runs,
    pause_workflow_batch,
    resume_workflow_batch,
    retry_failed_workflow_batch,
    retry_workflow_step,
    update_workflow_step_input,
)
from services.workflow_runtime import workflow_runtime

router = APIRouter(prefix="/workflows", tags=["workflows"])


class WorkflowDefinitionRequest(BaseModel):
    definition: dict = Field(default_factory=dict)


class WorkflowRunRequest(BaseModel):
    definition_key: str
    version: int = 0
    name: str = ""
    input: dict = Field(default_factory=dict)


class WorkflowBatchItemRequest(BaseModel):
    name: str = ""
    input: dict = Field(default_factory=dict)
    metadata: dict = Field(default_factory=dict)


class WorkflowBatchRequest(BaseModel):
    definition_key: str
    version: int = 0
    name: str = ""
    concurrency: int = 1
    items: list[WorkflowBatchItemRequest] = Field(default_factory=list)


class WorkflowStepInputRequest(BaseModel):
    input: dict = Field(default_factory=dict)


@router.get("/definitions")
def list_definitions(include_disabled: bool = False):
    return {"items": list_workflow_definitions(include_disabled=include_disabled)}


@router.get("/adapters")
def list_adapters():
    return {"items": list_workflow_adapters()}


@router.post("/definitions")
def upsert_definition(body: WorkflowDefinitionRequest):
    try:
        return create_or_update_workflow_definition(body.definition)
    except ValueError as exc:
        raise HTTPException(400, str(exc)) from exc


@router.get("/batches")
def list_batches(limit: int = 50, offset: int = 0, status: str = "", definition_key: str = ""):
    return list_workflow_batches(limit=limit, offset=offset, status=status, definition_key=definition_key)


@router.get("/batches/{batch_id}")
def get_batch(batch_id: str):
    batch = get_workflow_batch(batch_id)
    if not batch:
        raise HTTPException(404, "工作流批次不存在")
    return batch


@router.get("/batches/{batch_id}/summary")
def get_batch_summary(batch_id: str):
    summary = get_workflow_batch_summary(batch_id)
    if not summary:
        raise HTTPException(404, "工作流批次不存在")
    return summary


@router.post("/batches/{batch_id}/pause")
def pause_batch(batch_id: str):
    batch = pause_workflow_batch(batch_id)
    if not batch:
        raise HTTPException(404, "工作流批次不存在")
    return batch


@router.post("/batches/{batch_id}/resume")
def resume_batch(batch_id: str):
    batch = resume_workflow_batch(batch_id)
    if not batch:
        raise HTTPException(404, "工作流批次不存在")
    workflow_runtime.wake_up()
    return batch


@router.post("/batches/{batch_id}/cancel")
def cancel_batch(batch_id: str):
    batch = cancel_workflow_batch(batch_id)
    if not batch:
        raise HTTPException(404, "工作流批次不存在")
    workflow_runtime.wake_up()
    return batch


@router.post("/batches/{batch_id}/retry-failed")
def retry_failed_batch(batch_id: str):
    batch = retry_failed_workflow_batch(batch_id)
    if not batch:
        raise HTTPException(404, "工作流批次不存在")
    workflow_runtime.wake_up()
    return batch


@router.get("/runs")
def list_runs(limit: int = 50, offset: int = 0, status: str = "", definition_key: str = "", batch_id: str = ""):
    return list_workflow_runs(
        limit=limit,
        offset=offset,
        status=status,
        definition_key=definition_key,
        batch_id=batch_id,
    )


@router.post("/runs")
def start_run(body: WorkflowRunRequest):
    run = create_workflow_run(
        definition_key=body.definition_key,
        version=body.version,
        name=body.name,
        inputs=body.input,
    )
    if not run:
        raise HTTPException(404, "工作流定义不存在或未启用")
    workflow_runtime.wake_up()
    return run


@router.post("/runs/batch")
def start_batch(body: WorkflowBatchRequest):
    try:
        batch = create_workflow_batch(
            definition_key=body.definition_key,
            version=body.version,
            name=body.name,
            concurrency=body.concurrency,
            items=[item.model_dump() for item in body.items],
        )
    except ValueError as exc:
        raise HTTPException(400, str(exc)) from exc
    if not batch:
        raise HTTPException(404, "工作流定义不存在或未启用")
    workflow_runtime.wake_up()
    return batch


@router.get("/runs/{run_id}/summary")
def get_run_summary(run_id: str):
    summary = get_workflow_run_summary(run_id)
    if not summary:
        raise HTTPException(404, "工作流不存在")
    return summary


@router.get("/runs/{run_id}")
def get_run(run_id: str):
    run = get_workflow_run(run_id)
    if not run:
        raise HTTPException(404, "工作流不存在")
    return run


@router.get("/runs/{run_id}/events")
def get_events(
    run_id: str,
    since: int = 0,
    before: int = 0,
    limit: int = 200,
    latest: bool = False,
):
    events = list_workflow_events(run_id, since=since, before=before, limit=limit, latest=latest)
    if events is None:
        raise HTTPException(404, "工作流不存在")
    return events


@router.post("/runs/{run_id}/cancel")
def cancel_run(run_id: str):
    run = cancel_workflow_run(run_id)
    if not run:
        raise HTTPException(404, "工作流不存在")
    workflow_runtime.wake_up()
    return run


@router.post("/runs/{run_id}/steps/{step_id}/retry")
def retry_step(run_id: str, step_id: str):
    try:
        run = retry_workflow_step(run_id, step_id)
    except ValueError as exc:
        raise HTTPException(400, str(exc)) from exc
    if not run:
        raise HTTPException(404, "工作流或步骤不存在")
    workflow_runtime.wake_up()
    return run


@router.patch("/runs/{run_id}/steps/{step_id}/input")
def update_step_input(run_id: str, step_id: str, body: WorkflowStepInputRequest):
    try:
        run = update_workflow_step_input(run_id, step_id, body.input)
    except ValueError as exc:
        raise HTTPException(400, str(exc)) from exc
    if not run:
        raise HTTPException(404, "工作流或步骤不存在")
    return run

File: frontend/src/lib/workflows.ts
import { translate, type Language } from '@/lib/i18n'
import { apiFetch } from '@/lib/utils'

export type WorkflowDefinition = {
  id: number
  key: string
  version: number
  name: string
  description: string
  enabled: boolean
  definition: {
    steps?: WorkflowDefinitionStep[]
    sample_input?: Record<string, unknown>
    input_schema?: Record<string, unknown>
    ui_schema?: WorkflowUiSchema
    [key: string]: unknown
  }
  created_at?: string
  updated_at?: string
}

export type WorkflowUiOption = {
  label: string
  value: string | number | boolean
}

export type WorkflowUiField = {
  path: string
  label: string
  type?: 'text' | 'number' | 'select' | 'boolean'
  advanced?: boolean
  options?: WorkflowUiOption[]
  placeholder?: string
  helper?: string
  min?: number
  max?: number
}

export type WorkflowUiSection = {
  title: string
  description?: string
  fields?: WorkflowUiField[]
}

export type WorkflowUiSchema = {
  sections?: WorkflowUiSection[]
}

export type WorkflowAdapter = {
  key: string
}

export type WorkflowDefinitionStep = {
  id: string
  name?: string
  uses: string
  needs?: string[]
  timeout?: string
  max_attempts?: number
}

export type WorkflowStepRun = {
  id: string
  workflow_run_id: string
  step_id: string
  name: string
  adapter_key: string
  status: string
  attempt: number
  max_attempts: number
  input: Record<string, unknown>
  output: Record<string, unknown>
  error: Record<string, unknown>
  external_ref: string
  idempotency_key: string
  next_run_at?: string
  timeout_at?: string
  started_at?: string
  finished_at?: string
  duration_seconds?: number
  created_at?: string
  updated_at?: string
}

export type WorkflowRun = {
  id: string
  definition_key: string
  definition_version: number
  name: string
  status: string
  terminal: boolean
  input: Record<string, unknown>
  context?: Record<string, unknown>
  output?: Record<string, unknown>
  metadata?: Record<string, unknown>
  definition?: Record<string, unknown>
  batch_id?: string
  batch_item_index?: number
  current_step_id: string
  error: string
  cancellation_requested_at?: string
  started_at?: string
  finished_at?: string
  created_at?: string
  updated_at?: string
  steps?: WorkflowStepRun[]
}

export type WorkflowStatusSummary = {
  total: number
  pending: number
  running: number
  waiting_external: number
  retry_scheduled: number
  needs_attention: number
  cancel_requested: number
  succeeded: number
  failed: number
  cancelled: number
  terminal: number
  active: number
}

export type WorkflowObservability = {
  duration_seconds_avg: number
  duration_seconds_max: number
  stuck: number
}

export type WorkflowBatch = {
  id: string
  definition_key: string
  definition_version: number
  name: string
  status: string
  terminal: boolean
  total: number
  concurrency: number
  input: Record<string, unknown>
  summary: WorkflowStatusSummary
  observability?: WorkflowObservability
  runs?: WorkflowRun[] | null
  created_at?: string
  updated_at?: string
}

export type WorkflowStepSummary = {
  step_id: string
  name: string
  status: string
  adapter_key: string
  attempt: number
  max_attempts: number
  error_code: string
  error_message: string
  error_category: string
  operator_hint: string
  external_ref: string
  duration_seconds: number
  stuck: boolean
  stuck_reason: string
}

export type WorkflowRunSummary = {
  run_id: string
  batch_id?: string
  batch_item_index?: number
  definition_key: string
  status: string
  terminal: boolean
  account_id: number
  email: string
  current_stage: string
  display_status: string
  operator_action: string
  risk: string
  duration_seconds: number
  stuck: boolean
  stuck_reason: string
  stuck_step_id: string
  steps: WorkflowStepSummary[]
}

export type WorkflowBatchSummary = {
  id: string
  definition_key: string
  definition_version: number
  name: string
  status: string
  terminal: boolean
  total: number
  concurrency: number
  summary: WorkflowStatusSummary
  observability?: WorkflowObservability
  runs: WorkflowRunSummary[]
  created_at?: string
  updated_at?: string
}

export type WorkflowEvent = {
  id: number
  workflow_run_id: string
  step_id: string
  type: string
  level: string
  message: string
  line: string
  detail: Record<string, unknown>
  created_at?: string
}

export const WORKFLOW_STATUS_VARIANTS: Record<string, 'default' | 'success' | 'warning' | 'danger' | 'secondary'> = {
  pending: 'secondary',
  ready: 'secondary',
  running: 'default',
  waiting_external: 'warning',
  retry_scheduled: 'warning',
  needs_attention: 'warning',
  cancel_requested: 'warning',
  paused: 'warning',
  succeeded: 'success',
  failed: 'danger',
  skipped: 'secondary',
  cancelled: 'warning',
}

export const CANCELLABLE_WORKFLOW_STATUSES = new Set([
  'pending',
  'running',
  'waiting_external',
  'retry_scheduled',
  'needs_attention',
  'cancel_requested',
])

export const RETRYABLE_STEP_STATUSES = new Set([
  'failed',
  'needs_attention',
  'cancelled',
])

export function getWorkflowStatusText(status: string, language?: Language) {
  switch (status) {
    case 'pending':
      return translate('workflowStatus.pending', language)
    case 'ready':
      return translate('workflowStatus.ready', language)
    case 'running':
      return translate('workflowStatus.running', language)
    case 'waiting_external':
      return translate('workflowStatus.waiting_external', language)
    case 'retry_scheduled':
      return translate('workflowStatus.retry_scheduled', language)
    case 'needs_attention':
      return translate('workflowStatus.needs_attention', language)
    case 'cancel_requested':
      return translate('workflowStatus.cancel_requested', language)
    case 'paused':
      return translate('workflowStatus.paused', language)
    case 'succeeded':
      return translate('workflowStatus.succeeded', language)
    case 'failed':
      return translate('workflowStatus.failed', language)
    case 'skipped':
      return translate('workflowStatus.skipped', language)
    case 'cancelled':
      return translate('workflowStatus.cancelled', language)
    default:
      return status || '-'
  }
}

export async function fetchWorkflowDefinitions() {
  const data = await apiFetch('/workflows/definitions')
  return (data.items || []) as WorkflowDefinition[]
}

export async function fetchWorkflowAdapters() {
  const data = await apiFetch('/workflows/adapters')
  return (data.items || []) as WorkflowAdapter[]
}

export async function saveWorkflowDefinition(definition: Record<string, unknown>) {
  return apiFetch('/workflows/definitions', {
    method: 'POST',
    body: JSON.stringify({ definition }),
  }) as Promise<WorkflowDefinition>
}

export async function fetchWorkflowRuns(params: {
  limit: number
  offset: number
  status?: string
  definition_key?: string
  batch_id?: string
}) {
  const search = new URLSearchParams({
    limit: String(params.limit),
    offset: String(params.offset),
  })
  if (params.status) search.set('status', params.status)
  if (params.definition_key) search.set('definition_key', params.definition_key)
  if (params.batch_id) search.set('batch_id', params.batch_id)
  return apiFetch(`/workflows/runs?${search.toString()}`) as Promise<{
    items: WorkflowRun[]
    total: number
    running: number
    limit: number
    offset: number
  }>
}

export async function fetchWorkflowBatches(params: {
  limit: number
  offset: number
  status?: string
  definition_key?: string
}) {
  const search = new URLSearchParams({
    limit: String(params.limit),
    offset: String(params.offset),
  })
  if (params.status) search.set('status', params.status)
  if (params.definition_key) search.set('definition_key', params.definition_key)
  return apiFetch(`/workflows/batches?${search.toString()}`) as Promise<{
    items: WorkflowBatch[]
    total: number
    limit: number
    offset: number
  }>
}

export async function createWorkflowRun(payload: {
  definition_key: string
  version?: number
  name?: string
  input: Record<string, unknown>
}) {
  return apiFetch('/workflows/runs', {
    method: 'POST',
    body: JSON.stringify(payload),
  }) as Promise<WorkflowRun>
}

export async function createWorkflowBatch(payload: {
  definition_key: string
  version?: number
  name?: string
  concurrency?: number
  items: Array<{
    name?: string
    input: Record<string, unknown>
    metadata?: Record<string, unknown>
  }>
}) {
  return apiFetch('/workflows/runs/batch', {
    method: 'POST',
    body: JSON.stringify(payload),
  }) as Promise<WorkflowBatch>
}

export async function fetchWorkflowRun(runId: string) {
  return apiFetch(`/workflows/runs/${runId}`) as Promise<WorkflowRun>
}

export async function fetchWorkflowBatch(batchId: string) {
  return apiFetch(`/workflows/batches/${batchId}`) as Promise<WorkflowBatch>
}

export async function fetchWorkflowRunSummary(runId: string) {
  return apiFetch(`/workflows/runs/${runId}/summary`) as Promise<WorkflowRunSummary>
}

export async function fetchWorkflowBatchSummary(batchId: string) {
  return apiFetch(`/workflows/batches/${batchId}/summary`) as Promise<WorkflowBatchSummary>
}

export async function pauseWorkflowBatch(batchId: string) {
  return apiFetch(`/workflows/batches/${batchId}/pause`, { method: 'POST' }) as Promise<WorkflowBatch>
}

export async function resumeWorkflowBatch(batchId: string) {
  return apiFetch(`/workflows/batches/${batchId}/resume`, { method: 'POST' }) as Promise<WorkflowBatch>
}

export async function cancelWorkflowBatch(batchId: string) {
  return apiFetch(`/workflows/batches/${batchId}/cancel`, { method: 'POST' }) as Promise<WorkflowBatchSummary>
}

export async function retryFailedWorkflowBatch(batchId: string) {
  return apiFetch(`/workflows/batches/${batchId}/retry-failed`, { method: 'POST' }) as Promise<WorkflowBatchSummary & { retried?: number }>
}

export async function fetchWorkflowEvents(runId: string, since = 0) {
  const search = new URLSearchParams({ since: String(since), limit: '200' })
  const data = await apiFetch(`/workflows/runs/${runId}/events?${search.toString()}`)
  return data as { items: WorkflowEvent[]; cursor: number }
}

export async function cancelWorkflowRun(runId: string) {
  return apiFetch(`/workflows/runs/${runId}/cancel`, { method: 'POST' }) as Promise<WorkflowRun>
}

export async function retryWorkflowStep(runId: string, stepId: string) {
  return apiFetch(`/workflows/runs/${runId}/steps/${stepId}/retry`, { method: 'POST' }) as Promise<WorkflowRun>
}

export async function updateWorkflowStepInput(
  runId: string,
  stepId: string,
  input: Record<string, unknown>,
) {
  return apiFetch(`/workflows/runs/${runId}/steps/${stepId}/input`, {
    method: 'PATCH',
    body: JSON.stringify({ input }),
  }) as Promise<WorkflowRun>
}
